### PR TITLE
fix(meta): refuse to start cluster if data directory is used by another instance

### PR DESCRIPTION
**Pull request by Gun9niR**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Reused the `TrackingId` from telemetry as `ClusterId`. Its presence/absence in meta store is used to determine if the cluster is being created. If so, the meta node will write a `cluster_id/0` file in the specified data directory, signifying its occupancy. If this file is present on cluster creation, the cluster will refuse to start. 

Defer system parameter persistence to before starting RPC services, so that if invalid parameters prevent the meta node from starting, they will not be persisted.

The problem of reusing `TrackingId` is that backward compatibility is broken because its cf is changed.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment

### Release note

The object store URL identified by state_store and data_directory must not be shared by multiple clusters, or later clusters will refuse to start.

### CODE REVIEW

File: src/meta/src/hummock/manager/mod.rs [hunk 1]
```diff
@@ -2265,3 +2280,29 @@ fn gen_version_delta<'a>(
...
+
+    if let Ok(cluster_id) = cluster_id_res {
+        Err(ObjectError::internal(format!(
+            "data directory is already used by another cluster with id {:?}",
+            String::from_utf8(cluster_id.to_vec()).unwrap()
+        ))
+        .into())
+    } else {
+        // FIXME: Can't distinguish no such item from other errors.
+        object_store
```
Comment (zwang28) [inline src/meta/src/hummock/manager/mod.rs:2300]: So before this is fixed, a temporary object store read error may result in unexpected cluster id object overwriting, which is risky.

Comment (Gun9niR) [reply src/meta/src/hummock/manager/mod.rs:2300]: I tried to use `list`, but opendal requires the path to be a directory, which behaves differently from s3.

Comment (Gun9niR) [reply src/meta/src/hummock/manager/mod.rs]: I have put the file in another directory called `cluster_id` and uses list to check the file's existence.

**Review by zwang28 — APPROVED** [review]
LGTM.
We may replace the list trick after refactoring object store's get error.

File: src/meta/src/telemetry.rs [hunk 2]
```diff
@@ -14,99 +14,17 @@
...
-use uuid::Uuid;
 
 use crate::manager::ClusterManager;
-use crate::model::{MetadataModelError, MetadataModelResult};
-use crate::storage::{MetaStore, Snapshot};
-
-/// Column in meta store
-pub const TELEMETRY_CF: &str = "cf/telemetry";
-/// `telemetry` in bytes
-pub const TELEMETRY_KEY: &[u8] = &[74, 65, 0x6c, 65, 0x6d, 65, 74, 72, 79];
```
Comment (odysa) [inline src/meta/src/telemetry.rs:34]: The only concern is that clusters may lose the previous persistent `tracking_id` after upgradation. Not a big problem.

**Review by odysa — COMMENTED** [review]
LGTM

**Review by odysa — APPROVED** [review]

**Comment by hzxa21** [comment]
> The problem of reusing TrackingId is that backward compatibility is broken because its cf is changed.

What happen when upgrading the cluster deployed prior to this PR? Will the cluster fail to startup because cluster id is not found in the new CF?

**Comment by Gun9niR** [comment]
If the cluster id is not found, meta node will consider the cluster to be launched for the first time, resulting in overriding existing system parameters.

In addition, since the id is newly generated, telemetry's report id will be different.

**Comment by hzxa21** [comment]
> If the cluster id is not found, meta node will consider the cluster to be launched for the first time, resulting in overriding existing system parameters.
> 
> In addition, since the id is newly generated, telemetry's report id will be different.

I see. As long as the cluster can start up without any issue, creating a new telemetry report id sounds okay to me.

**Comment by Gun9niR** [comment]
> overriding existing system parameters.

Though I'm more worried about this, since some parameters cannot be modified, the override will be irreversible.

**Pull request merged**